### PR TITLE
Bundler: Use RuboCop 0.80.1, rubocop-performance 1.5.2

### PR DESCRIPTION
**Pull request by olleolleolle**
# Description:

- Bundler: use the newest RuboCop, `0.80.1`
- Bundler: use the newest `rubocop-performance`, `1.5.2`
- GitHub Actions: use `rubocop` `0.80.1` for linting
- Add Rake tasks to run RuboCop for rubygems, bundler

A follow-up to  https://github.com/rubygems/rubygems/pull/3413#issuecomment-602538004

(As a step after this, the RuboCop configurations can be unified, and the codestyle be one. None of that will happen in this PR.)

______________

# Tasks:

- [x] Describe the problem / feature
- [ ] Write tests
- [x] Write code to solve the problem
- [x] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: .github/workflows/ubuntu-lint.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ jobs:
         run: rake setup
       - name: Run Lint
         run: |
-          gem install rubocop -v "~>0.74.0"
+          gem install rubocop -v "~>0.80.1"
```
Comment (deivid-rodriguez) [inline .github/workflows/ubuntu-lint.yml]: Can you verify whether we actually need this? I'd say we don't because `rake setup` should take care of it.

File: .github/workflows/ubuntu-lint.yml [hunk 2]
```diff
@@ -21,7 +21,7 @@ jobs:
...
           util/rubocop
       - name: Install Dependencies in Bundler
```
Comment (deivid-rodriguez) [inline .github/workflows/ubuntu-lint.yml]: If we try my suggestion, the `rm ../Gemfile*` below should be removed.

**Comment by olleolleolle** [comment]
@deivid-rodriguez Thanks for the support!

> `util/rubocop --config .rubocop.yml bundler`

That led me to change the excludes, so that I could run it.

```diff
diff --git a/.rubocop.yml b/.rubocop.yml
index 1087db19c..2db882e03 100644
--- a/.rubocop.yml
+++ b/.rubocop.yml
@@ -1,7 +1,7 @@
 AllCops:
   DisabledByDefault: true
   Exclude:
-    - 'bundler/**/*'
+    - 'bundler/lib/bundler/vendor/**/*'
     - 'lib/rubygems/resolver/molinillo/**/*'
     - 'pkg/**/*'
     - 'tmp/**/*'
```

That runs, and:

> 371 files inspected, 584 offenses detected

**Comment by deivid-rodriguez** [comment]
Sorry, I mistyped the command, it should be `util/rubocop --config bundler/.rubocop.yml bundler` and I don't think you need to change any configuration.

File: .github/workflows/ubuntu-lint.yml [hunk 3]
```diff
@@ -21,15 +21,14 @@ jobs:
...
-          util/rubocop
+          rake rubocop:rubygems
       - name: Install Dependencies in Bundler
         run: |
           bin/rake spec:deps
         working-directory: ./bundler
       - name: Run Lint in Bundler
         run: |
-          bin/rubocop --ignore_parent_exclusion
+          rake rubocop:bundler
```
Comment (deivid-rodriguez) [inline .github/workflows/ubuntu-lint.yml]: You should move this to a separate step so that `working-directory: ./bundler` does not apply.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -11,7 +11,8 @@ contributors to follow to reduce the time it takes to get changes merged in.
 2.  Ensure that your code blends well with ours:
     *   No trailing whitespace
     *   Match indentation (two spaces)
-    *   Match coding style (run `rake rubocop`)
+    *   Match coding style (run `rake rubocop:rubygems` or
+        `rake rubocop:bundler`)
```
Comment (deivid-rodriguez) [inline CONTRIBUTING.md]: Can we instead make `rake rubocop` run both tasks, so we don't have to change this and we run lints for both repos by default. It's possible that PRs will change both repos, and I think it will get more common as we unify the libraries more and more.

Comment (olleolleolle) [reply CONTRIBUTING.md]: Nice, that'd be even better.

**Comment by olleolleolle** [comment]
@deivid-rodriguez Perhaps we need `rake` in place of the rubocop in the Bundler thing?

```

Run bin/rake check_rvm_integration man:check
  bin/rake check_rvm_integration man:check
  shell: /bin/bash -e {0}
  env:
    PATH: /home/runner/.rubies/ruby-2.6.5/bin:/usr/share/rust/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/home/runner/.dotnet/tools:/home/runner/.config/composer/vendor/bin
gem install rubygems-bundler && RUBYOPT=-Ilib rake -T
Building native extensions. This could take a while...
# In case of problems run the following command to update binstubs:
    gem regenerate_binstubs
Successfully installed executable-hooks-1.6.0
Successfully installed bundler-unload-1.0.2
Successfully installed rubygems-bundler-1.4.5
Parsing documentation for executable-hooks-1.6.0
Installing ri documentation for executable-hooks-1.6.0
Parsing documentation for bundler-unload-1.0.2
Installing ri documentation for bundler-unload-1.0.2
Parsing documentation for rubygems-bundler-1.4.5
Installing ri documentation for rubygems-bundler-1.4.5
Done installing documentation for executable-hooks, bundler-unload, rubygems-bundler after 0 seconds
3 gems installed
/home/runner/.rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/exe/rake:25:in `require': cannot load such file -- rake (LoadError)
	from /home/runner/.rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/exe/rake:25:in `<top (required)>'
	from /home/runner/.rubies/ruby-2.6.5/bin/rake:23:in `load'
	from /home/runner/.rubies/ruby-2.6.5/bin/rake:23:in `<main>'
	from /home/runner/.rubies/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
	from /home/runner/.rubies/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
rake aborted!

```

**Comment by deivid-rodriguez** [comment]
Yeah, maybe... I don't fully understand the issue but yeah, maybe that will work. Or maybe change the `rvm_integration` task to be `gem install rubygems-bundler rake && RUBYOPT=-Ilib rake -T`.

**Comment by olleolleolle** [comment]
That did it!

File: .github/workflows/ubuntu-lint.yml [hunk 5]
```diff
@@ -19,18 +19,14 @@ jobs:
           ruby-version: 2.6.5
       - name: Install Dependencies
         run: rake setup
-      - name: Run Lint
-        run: |
-          gem install rubocop -v "~>0.74.0"
-          util/rubocop
```
Comment (deivid-rodriguez) [inline .github/workflows/ubuntu-lint.yml]: Maybe keep the "Run Lint" title and run `rake rubocop` here, and we don't need the extra step running `rake rubocop:bundler`.

File: Rakefile [hunk 6]
```diff
@@ -64,11 +64,20 @@ rescue LoadError
...
-  sh "util/rubocop"
+namespace :rubocop do
+  desc "Run rubocop for RubyGems"
+  task(:rubygems) do
+    sh "util/rubocop"
+  end
+
+  desc "Run rubocop for Bundler"
+  task(:bundler) do
+    sh "util/rubocop --ignore_parent_exclusion --config bundler/.rubocop.yml bundler"
```
Comment (deivid-rodriguez) [inline Rakefile]: So, is the `--ignore_parent_exclusion` flag actually needed? I think we no longer need it since we are running `rubocop` from the root folder.

Comment (olleolleolle) [reply Rakefile]: Let's attempt without it.

File: .github/workflows/ubuntu-lint.yml (hunk 5 above)
Comment (olleolleolle) [reply .github/workflows/ubuntu-lint.yml]: Thanks, this is getting simpler each time!

**Comment by olleolleolle** [comment]
Now also enabled "--parallel | Use available CPUs to execute inspection in parallel. "

**Comment by olleolleolle** [comment]
@deivid-rodriguez How do you feel about this kind of addition?

```diff
diff --git a/Rakefile b/Rakefile
index 6ccca5c62..3297d286f 100644
--- a/Rakefile
+++ b/Rakefile
@@ -70,9 +70,13 @@ namespace :rubocop do
     sh "util/rubocop --parallel"
   end
 
-  desc "Run rubocop for Bundler"
-  task(:bundler) do
-    sh "util/rubocop --parallel --config bundler/.rubocop.yml bundler"
+  desc "Run rubocop for Bundler. Set auto_correct to 'auto' to auto-correct."
+  task(:bundler, [:auto_correct]) do |t, args|
+    if args[:auto_correct] == 'auto'
+      sh "util/rubocop -a --config bundler/.rubocop.yml bundler"
+    else
+      sh "util/rubocop --parallel --config bundler/.rubocop.yml bundler"
+    end
   end
 end
 
diff --git a/bundler/Rakefile b/bundler/Rakefile
index fbbc19c07..42cf66ed9 100644
--- a/bundler/Rakefile
+++ b/bundler/Rakefile
@@ -78,11 +78,6 @@ namespace :spec do
   end
 end
 
-desc "Run RuboCop"
-task :rubocop do
-  sh("bin/rubocop --parallel")
-end
-
 desc "Check RVM integration"
 task :check_rvm_integration do
   # The rubygems-bundler gem is installed by RVM by default and it could easily
diff --git a/bundler/doc/development/PULL_REQUESTS.md b/bundler/doc/development/PULL_REQUESTS.md
index 09428f057..c2656b12f 100644
--- a/bundler/doc/development/PULL_REQUESTS.md
+++ b/bundler/doc/development/PULL_REQUESTS.md
@@ -10,7 +10,7 @@ Before you submit a pull request, please remember to do the following:
 
 Make sure the code formatting and styling adheres to the guidelines. We use RuboCop for this. Lack of formatting adherence will result in automatic GitHub Actions build failures.
 
-      $ bin/rubocop -a
+      $ rake 'rubocop:bundler[auto]'
 
 ## Tests
 
```

**Comment by deivid-rodriguez** [comment]
Since we're ditching the `bin/rubocop` binstub, we should definitely provide alternatives. What about something like:

```ruby
task :bundler do |_, args|
  sh "util/rubocop", "--config", "bundler/.rubocop.yml", *args, "bundler"
end
```

So that we can pass arbitrary flags, in particular, `rake rubocop:bundler[-a]`. And I guess we could add the same for the `rubygems` task?

Also I grepped for `bin/rubocop` and I found another usage in `bundler/task/git_hooks.rake`. @bronzdoc uses that stuff but it doesn't seem straightforward to migrate to the new workflow, so I'm unsure what to do.

@bronzdoc thoughts?

**Comment by bronzdoc** [comment]
@deivid-rodriguez @olleolleolle I use something similar for Rubygems but it doesn't use `bin/rubocop` 
The hooks in Rubygems are in `.githooks` and configured with:
```ruby
desc "Setup git hooks"
task :git_hooks do
  sh "git config core.hooksPath .githooks"
end
```

seems like `bundler/task/git_hooks.rake` is something that @segiddins created to do the same.

**Comment by olleolleolle** [comment]
[`bundler/spec/quality_spec.rb`](https://github.com/rubygems/rubygems/blob/master/bundler/spec/quality_spec.rb#L5) seems to be the net new to what we have looked at in this PR.

Checks for loose ends, like conflict markers staying, etc.

**Comment by deivid-rodriguez** [comment]
An idea could be to keep the `bin/rubocop` binstub and do something like this:

```bash
#!/usr/bin/env bash

cd "${BASH_SOURCE[0]%/*}/../.." || exit 1

util/rubocop --config bundler/.rubocop.yml "$@" bundler
```

And shell out to that from `rake rubocop:bundler`.

Then we could keep all the existing documentation using `bin/rubocop` and the git hook should be straighforward to modify to keep working (run it directly instead of through `ruby` I guess).

**Comment by deivid-rodriguez** [comment]
> seems like bundler/task/git_hooks.rake is something that @segiddins created to do the same.

Oh, I see. Then I guess another alternative would be to kill those for now (since @segiddins is no longer involved), and later you can iterate going forward to suite your git hook needs?

**Comment by olleolleolle** [comment]
I sketched in a diff

```diff
diff --git a/.githooks/pre-commit b/.githooks/pre-commit
index 6ac7eb6c3..d4406cbbb 100755
--- a/.githooks/pre-commit
+++ b/.githooks/pre-commit
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-HOOKS="pre-commit-rubocop"
+HOOKS="pre-commit-rubocop precommit-bundler-quality"
 
 for hook in $HOOKS; do
   if [ -f "$PWD/.githooks/$hook" ]; then
diff --git a/.githooks/pre-commit-bundler-quality b/.githooks/pre-commit-bundler-quality
new file mode 100644
index 000000000..942eb13e4
--- /dev/null
+++ b/.githooks/pre-commit-bundler-quality
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+files=$(git diff --name-only --cached --diff-filter=d)
+length=${#files}
+
+# Run Bundler quality hook only when analyzable files are present
+if [[ $length -ne 0 ]]; then
+  pushd bundler
+  rspec spec/quality_spec.rb
+  popd
+else
+  echo "No files to analyze"
+fi
diff --git a/.githooks/pre-commit-rubocop b/.githooks/pre-commit-rubocop
index 87fab94c1..50f1670df 100755
--- a/.githooks/pre-commit-rubocop
+++ b/.githooks/pre-commit-rubocop
@@ -5,7 +5,7 @@ length=${#files}
 
 # Run rubocop hook only when analyzable files are present
 if [[ $length -ne 0 ]]; then
-  ./util/rubocop $files
+  ./util/rubocop --parallel $files
 else
   echo "No files to analyze"
 fi
```

**Comment by deivid-rodriguez** [comment]
@olleolleolle I'm not sure about your last diff (you'll need to check with @bronzdoc since he's the only maintainer using git hooks at the moment I think?). In any case, it seems like that's starting to fall a bit out of scope?

**Comment by olleolleolle** [comment]
@deivid-rodriguez Let's go with "remove the githook files in bundler" instead. It's a way forward.

**Comment by deivid-rodriguez** [comment]
I'm good with that, yeah :+1:.

**Comment by deivid-rodriguez** [comment]
What do you think of https://github.com/rubygems/rubygems/pull/3416#issuecomment-602755130, though? I think it will allow more easily running `rubocop` from within the `bundler` subfolder (without any changes to how it's been run until today). In my case, it will allow my editor to autorun rubocop when I make syntax changes, because it automatically looks for a `bin/rubocop` binstub.

**Comment by olleolleolle** [comment]
@deivid-rodriguez I edited my commit history, amended in changes we talked about.

Take a look, and see if the `bundler/bin/rubocop` was what you meant.

File: bundler/doc/development/PULL_REQUESTS.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ Before you submit a pull request, please remember to do the following:
 
 Make sure the code formatting and styling adheres to the guidelines. We use RuboCop for this. Lack of formatting adherence will result in automatic GitHub Actions build failures.
 
-      $ bin/rubocop -a
+      $ rake 'rubocop:bundler[-a]'
```
Comment (deivid-rodriguez) [inline bundler/doc/development/PULL_REQUESTS.md]: This documentation file assumes you're in the bundler folder, so the new version will no longer work. Since we have now provided a working `bin/rubocop` binstub, we can revert this change.

Comment (olleolleolle) [reply bundler/doc/development/PULL_REQUESTS.md]: Dropped the change.

**Comment by olleolleolle** [comment]
I have updated (edited out of the commit history).

**Comment by deivid-rodriguez** [comment]
Thanks @olleolleolle!
